In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
import warnings
warnings.filterwarnings ('ignore')

In [2]:
AW_data= pd.read_csv('C:/Users/user/Desktop/Assignment Weather AUSTRALIA/weatherAUS.csv')
AW_test_data= pd.read_csv('C:/Users/user/Desktop/Assignment Weather AUSTRALIA/test.csv')

In [3]:
AW_data.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  12/1/2008   Albury     13.4     22.9       0.6          NaN       NaN   
1  12/2/2008   Albury      7.4     25.1       0.0          NaN       NaN   
2  12/3/2008   Albury     12.9     25.7       0.0          NaN       NaN   
3  12/4/2008   Albury      9.2     28.0       0.0          NaN       NaN   
4  12/5/2008   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

In [4]:
AW_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           140863 non-null  object 
 1   Location       140863 non-null  object 
 2   MinTemp        139383 non-null  float64
 3   MaxTemp        139607 non-null  float64
 4   Rainfall       137614 non-null  float64
 5   Evaporation    78493 non-null   float64
 6   Sunshine       71726 non-null   float64
 7   WindGustDir    130602 non-null  object 
 8   WindGustSpeed  130665 non-null  float64
 9   WindDir9am     130575 non-null  object 
 10  WindDir3pm     136643 non-null  object 
 11  WindSpeed9am   139101 non-null  float64
 12  WindSpeed3pm   137807 non-null  float64
 13  Humidity9am    138215 non-null  float64
 14  Humidity3pm    136365 non-null  float64
 15  Pressure9am    125976 non-null  float64
 16  Pressure3pm    126015 non-null  float64
 17  Cloud9am       85379 non-null

In [5]:
AW_data.isnull().sum()

Date              4597
Location          4597
MinTemp           6077
MaxTemp           5853
Rainfall          7846
Evaporation      66967
Sunshine         73734
WindGustDir      14858
WindGustSpeed    14795
WindDir9am       14885
WindDir3pm        8817
WindSpeed9am      6359
WindSpeed3pm      7653
Humidity9am       7245
Humidity3pm       9095
Pressure9am      19484
Pressure3pm      19445
Cloud9am         60081
Cloud3pm         63517
Temp9am           6362
Temp3pm           8200
RainToday         7846
RainTomorrow      7851
dtype: int64

In [6]:
#using the label encoder
le=LabelEncoder()

In [7]:
cat_col = [i for i in AW_data.columns if AW_data[i].dtype == "object"]

In [8]:
#converting the categorical columns to numerical data using label encoder
for col in cat_col:
    AW_data[col] = le.fit_transform(AW_data[col])

In [9]:
print(AW_data)

        Date  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0        890         2     13.4     22.9       0.6          NaN       NaN   
1        989         2      7.4     25.1       0.0          NaN       NaN   
2       1088         2     12.9     25.7       0.0          NaN       NaN   
3       1115         2      9.2     28.0       0.0          NaN       NaN   
4       1124         2     17.5     32.3       1.0          NaN       NaN   
...      ...       ...      ...      ...       ...          ...       ...   
145455  2452        40      2.8     23.4       0.0          NaN       NaN   
145456  2462        40      3.6     25.3       0.0          NaN       NaN   
145457  2472        40      5.4     26.9       0.0          NaN       NaN   
145458  2482        40      7.8     27.0       0.0          NaN       NaN   
145459  2492        40     14.9      NaN       0.0          NaN       NaN   

        WindGustDir  WindGustSpeed  WindDir9am  ...  Humidity9am  Humidity3

In [10]:
AW_data_filled_mean = AW_data.fillna(AW_data.mean())

In [11]:
print(AW_data_filled_mean)

        Date  Location  MinTemp    MaxTemp  Rainfall  Evaporation  Sunshine  \
0        890         2     13.4  22.900000       0.6     5.327179    7.5322   
1        989         2      7.4  25.100000       0.0     5.327179    7.5322   
2       1088         2     12.9  25.700000       0.0     5.327179    7.5322   
3       1115         2      9.2  28.000000       0.0     5.327179    7.5322   
4       1124         2     17.5  32.300000       1.0     5.327179    7.5322   
...      ...       ...      ...        ...       ...          ...       ...   
145455  2452        40      2.8  23.400000       0.0     5.327179    7.5322   
145456  2462        40      3.6  25.300000       0.0     5.327179    7.5322   
145457  2472        40      5.4  26.900000       0.0     5.327179    7.5322   
145458  2482        40      7.8  27.000000       0.0     5.327179    7.5322   
145459  2492        40     14.9  23.002597       0.0     5.327179    7.5322   

        WindGustDir  WindGustSpeed  WindDir9am  ...

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X= AW_data_filled_mean.drop(columns=['RainTomorrow'])
y= AW_data_filled_mean['RainTomorrow']

In [14]:
#use train test split to split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

In [15]:
#Initialize the KNeighbours
model_KNN = KNeighborsClassifier()
model_KNN.fit(X_train, y_train)

KNeighborsClassifier()

In [16]:
#make predictions using the KNN classifier
y_predict_KNN= model_KNN.predict(X_test)
Accuracy_KNN= accuracy_score(y_test, y_predict_KNN)
print("Accuracy Score:")
print(Accuracy_KNN)

Accuracy Score:
0.8186786745497043


In [17]:
# Initialize the Decision Tree Classifier
model_DT = DecisionTreeClassifier()
model_DT.fit(X_train, y_train) #the train data for x and y are used to train

#make predictions
y_predict_DT = model_DT.predict(X_test)

Accuracy_DT = accuracy_score(y_test, y_predict_DT)
print("Accuracy Score:")
print(Accuracy_DT)

Accuracy Score:
0.7751615564416334


In [18]:
# Initialize and train RandomForestClassifier
model_rf = RandomForestClassifier(n_estimators=100, random_state=42)
model_rf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [19]:
# Predict on the test set
y_predict_rf = model_rf.predict(X_test)

Accuracy_RT= accuracy_score(y_test, y_predict_rf)
print("Accuracy score")
print(Accuracy_RT)

Accuracy score
0.8443558366561253


From the various models used for the prediction of the weather in Australia, it can be seen that the Random forest is preferable, followed by the KNN Classifier. These models gave a 84% and 81% accuracy score respoectively. 